server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -0,0 +1,165 @@
+import static org.elasticsearch.synonyms.SynonymsAPI.SYNONYMS_ORIGIN;
+import static org.elasticsearch.xcontent.XContentFactory.jsonBuilder;
+
+public class SynonymsManagementAPIService {
+    public static final String SYNONYMS_INDEX = ".synonyms";
+
+    public static final String SYNONYMS_FEATURE_NAME = "synonyms";
+    public static final String SYNONYMS_SET_FIELD = "synonyms_set";
+    public static final String SYNONYMS_FIELD = "synonyms";
+
+    private final Client client;
+
+    public static final SystemIndexDescriptor SYNONYMS_DESCRIPTOR = SystemIndexDescriptor.builder()
+        .setIndexPattern(SYNONYMS_INDEX + "*")
+        .setDescription("Synonyms index for synonyms managed through APIs")
+        .setPrimaryIndex(SYNONYMS_INDEX)
+        .setMappings(mappings())
+        .setSettings(settings())
+        .setVersionMetaKey("version")
+        .setOrigin(SYNONYMS_ORIGIN)
+        .build();
+
+    public SynonymsManagementAPIService(Client client) {
+        this.client = new OriginSettingClient(client, SYNONYMS_ORIGIN);
+    }
+
+    private static XContentBuilder mappings() {
+        try {
+            XContentBuilder builder = jsonBuilder();
+            builder.startObject();
+            {
+                builder.startObject(SINGLE_MAPPING_NAME);
+                {
+                    builder.startObject("_meta");
+                    {
+                        builder.field("version", Version.CURRENT.toString());
+                    }
+                    builder.endObject();
+                    builder.field("dynamic", "strict");
+                    builder.startObject("properties");
+                    {
+                        builder.startObject("synonyms");
+                        {
+                            builder.field("type", "text");
+                        }
+                        builder.endObject();
+                        builder.startObject(SYNONYMS_SET_FIELD);
+                        {
+                            builder.field("type", "keyword");
+                        }
+                        builder.endObject();
+                    }
+                    builder.endObject();
+                }
+                builder.endObject();
+            }
+            builder.endObject();
+            return builder;
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to build mappings for " + SYNONYMS_INDEX, e);
+        }
+    }
+
+    public void putSynonymSet(String resourceName, SynonymSet synonymSet, ActionListener<PutSynonymsAction.Response> listener) {
+
+        // TODO Add synonym rules validation
+
+        // Delete synonym set if it existed previously
+        DeleteByQueryRequest dbqRequest = new DeleteByQueryRequest(SYNONYMS_INDEX).setQuery(
+            QueryBuilders.termQuery(SYNONYMS_SET_FIELD, resourceName)
+        ).setIndicesOptions(IndicesOptions.fromOptions(true, true, false, false));
```

Comment:
Can you please explain why do we need these indices options?

Reply:
The idea is not to fail when the system index has not been created yet, as can happen if no synonym has been created.

That way, we don't have to catch the index not found exception, and can use the same code path for it.

I'll document it on the code to make it clear

Reply:
I'm afraid it is not created on startup - it's created when the first synonym set is added (checked that). I think we need those index options to simplify error management, but I'm open for suggestions!

Reply:
Thanks for explanation, @carlosdelest . I had a wrong assumption that the server system indices are created on node startup. Makes sense!